lightning-c-bindings/src/c_types/mod.rs

```diff
@@ -0,0 +1,415 @@
+		::bitcoin::blockdata::transaction::TxOut {
+			script_pubkey: self.script_pubkey.into_rust().into(),
+			value: self.value,
+		}
+	}
+	pub(crate) fn from_rust(txout: ::bitcoin::blockdata::transaction::TxOut) -> Self {
+		Self {
+			script_pubkey: CVecTempl::from(txout.script_pubkey.into_bytes()),
+			value: txout.value
+		}
+	}
+}
+#[no_mangle]
+pub extern "C" fn TxOut_free(_res: TxOut) { }
+
+#[repr(C)]
+pub struct u8slice {
+	pub data: *const u8,
+	pub datalen: usize
+}
+impl u8slice {
+	pub(crate) fn from_slice(s: &[u8]) -> Self {
+		Self {
+			data: s.as_ptr(),
+			datalen: s.len(),
+		}
+	}
+	pub(crate) fn to_slice(&self) -> &[u8] {
+		if self.datalen == 0 { return &[]; }
+		unsafe { std::slice::from_raw_parts(self.data, self.datalen) }
+	}
+}
+
+#[repr(C)]
+pub struct usizeslice {
+	pub data: *const usize,
+	pub datalen: usize
+}
+impl usizeslice {
+	pub(crate) fn from_slice(s: &[usize]) -> Self {
+		Self {
+			data: s.as_ptr(),
+			datalen: s.len(),
+		}
+	}
+	pub(crate) fn to_slice(&self) -> &[usize] {
+		if self.datalen == 0 { return &[]; }
+		unsafe { std::slice::from_raw_parts(self.data, self.datalen) }
+	}
+}
+
+#[repr(C)]
+#[derive(Copy, Clone)]
+/// Arbitrary 32 bytes, which could represent one of a few different things. You probably want to
+/// look up the corresponding function in rust-lightning's docs.
+pub struct ThirtyTwoBytes {
+	pub data: [u8; 32],
+}
+impl ThirtyTwoBytes {
+	pub(crate) fn empty() -> Self {
+		Self { data: [0; 32] }
+	}
+}
+
+#[repr(C)]
+pub struct ThreeBytes { pub data: [u8; 3], }
+#[derive(Clone)]
+#[repr(C)]
+pub struct FourBytes { pub data: [u8; 4], }
+#[derive(Clone)]
+#[repr(C)]
+pub struct TenBytes { pub data: [u8; 10], }
+#[derive(Clone)]
+#[repr(C)]
+pub struct SixteenBytes { pub data: [u8; 16], }
+
+pub(crate) struct VecWriter(pub Vec<u8>);
+impl lightning::util::ser::Writer for VecWriter {
+	fn write_all(&mut self, buf: &[u8]) -> Result<(), ::std::io::Error> {
+		self.0.extend_from_slice(buf);
+		Ok(())
+	}
+	fn size_hint(&mut self, size: usize) {
+		self.0.reserve_exact(size);
+	}
+}
+pub(crate) fn serialize_obj<I: lightning::util::ser::Writeable>(i: &I) -> derived::CVec_u8Z {
+	let mut out = VecWriter(Vec::new());
+	i.write(&mut out).unwrap();
+	CVecTempl::from(out.0)
+}
+pub(crate) fn deserialize_obj<I: lightning::util::ser::Readable>(s: u8slice) -> Result<I, lightning::ln::msgs::DecodeError> {
+	I::read(&mut s.to_slice())
+}
+
+#[repr(C)]
+#[derive(Copy, Clone)]
+/// A Rust str object, ie a reference to a UTF8-valid string.
+/// This is *not* null-terminated so cannot be used directly as a C string!
+pub struct Str {
+	pub chars: *const u8,
+	pub len: usize
+}
+impl Into<Str> for &'static str {
+	fn into(self) -> Str {
+		Str { chars: self.as_ptr(), len: self.len() }
+	}
+}
+impl Into<&'static str> for Str {
+	fn into(self) -> &'static str {
+		if self.len == 0 { return ""; }
+		std::str::from_utf8(unsafe { std::slice::from_raw_parts(self.chars, self.len) }).unwrap()
+	}
+}
+
```

Comment:
Maybe document that these *only* support static strs? 
because currently you could easily cause UB by doing:
```rust
let me = String::from("Hello World");
let s = Str { chars: me.as_ptr(), len: me.len()};
let new_me: &'static str = s.into();
```
Now you have a static reference to a temporary(owned String that will be deallocated at the end of the scope)
Now the fact that you have it isn't UB by it's own AFAIK but the second `me` is dropped, if you still have `new_me` now you have UB, because you have a dangling reference.

Reply:
Right, that kind of thing is kinda up to the generation code - certainly no Rust crates are going to be depending on this stuff so its only the C parts that form the "actual" public API. While its not used now, there's a strong possibility that in the future something like this happens:

```
fn calling_c(self: &TraitObj, s: &str) {
  (self.calling_c)(c_types::Str::from(s));
}
```

In general we have that pattern in a number of places - "reference" things are only valid for the stack that they exist on.